R/standardize_treatment_kernel.R

```diff
@@ -0,0 +1,427 @@
+                              as.numeric(nj), lambda, lowlim, uplim, data_in)
+
+  # create propensity score multipliers and split by site
+  pro_trt <- Z / (pscores * nj[S_factor])
+  pro_trt_split <- split(pro_trt, S_factor)
+  pro_ctr <- (1 - Z) / ((1 - pscores) * nj[S_factor])
+  pro_ctr_split <- split(pro_ctr, S_factor)
+
+  # construct linear term vector
+  if(verbose) message("Creating linear term vector...")
+  if(is.null(data_in$q)) {
+    q <- create_q_vector_treatment_kernel(Xtaus, pro_trt_split, Xtarget, kerneltau)
+  } else {
+    q <- data_in$q
+  }
+
+  # construct quadratic term matrix
+  if(verbose) message("Creating quadratic term matrix...")
+  if(is.null(data_in$P)) {
+    P <- create_P_matrix_treatment_kernel(n, X0s, Xtaus, kernel0, kerneltau,
+                                          pro_trt_split, pro_ctr_split)
+  } else {
+    P <- data_in$P
+  }
+  I0 <- create_I0_matrix_treatment(pro_trt_split, pro_ctr_split,
+                                   scale_sample_size, nj, n, 0)
+  P <- P + lambda * I0
+
+  # construct constraint matrix
+  if(verbose) message("Creating constraint matrix...")
+  if(is.null(data_in$constraints)) {
+    constraints <- create_constraints_treatment_kernel(X0s, Xtaus, Z, S_factor,
+                                                       pro_trt_split, pro_ctr_split,
+                                                       lowlim, uplim, verbose)
+  } else {
+    constraints <- data_in$constraints
+    constraints$l[(J + 1):(J + n)] <- lowlim
+    constraints$u[(J + 1):(J + n)] <- uplim
+  }
+
+  # set optimization settings
+  settings <- do.call(osqp::osqpSettings,
+                      c(list(verbose = verbose,
+                             eps_rel = eps_rel,
+                             eps_abs = eps_abs),
+                        list(...)))
+
+  # solve optimization problem (possibly with uniform weights)
+  if(init_uniform) {
+    if(verbose) message("Initializing with uniform weights")
+    unifw <- get_uniform_weights_treatment_kernel(nj)
+    obj <- osqp::osqp(P, q, constraints$A,
+                      constraints$l, constraints$u, pars = settings)
+    obj$WarmStart(x = unifw)
+    solution <- obj$Solve()
+  } else {
+    solution <- osqp::solve_osqp(P, q, constraints$A,
+                                 constraints$l, constraints$u,
+                                 pars = settings)
+  }
+
+  # convert weights into a matrix
+  if(verbose) message("Reordering weights...")
+  weights <- matrix(0, ncol = J, nrow = n)
+  cumsumnj <- cumsum(c(1, nj))
+  for(j in 1:J) {
+    weights[idxs[[j]], j] <- solution$x[cumsumnj[j]:(cumsumnj[j + 1] - 1)]
+  }
+
+  # compute imbalance matrix
+  imbalancetau <- as.matrix(colMeans(Xtarget) - t(Xtau) %*% (weights * pro_trt))
+  imbalance0 <- as.matrix(t(X0) %*% (weights * pro_trt) -
+                          t(X0) %*% (weights * pro_ctr))
+
+  # collapse weight matrix to vector
+  weights <- rowSums(weights)
```

Comment:
What do the weights sum to here? If they sum to 1 then the imbalance calculations should be fine, but if they sum to the number of units, then the imbalance calculations are comparing a mean and a sum

Reply:
The weights sum to the number of units, but I think (?) this is OK because the propensity score multipliers `pro_ctr`and `pro_trt` effectively divide by the number of units. As a simple example, if we have three units (X1, X2, X3) and weights (w1, w2, w3), then if we look at (w1X1 + w2X2 + w3X3) / 3, we want w1 + w2 + w3 = 3.